File: ios/Infrastructure/MLKit/Adapters/MLKitBarcodeAdapter.swift
import Foundation

#if MLKIT_BARCODE_SCANNING
  import MLKitBarcodeScanning

  class MLKitBarcodeAdapter {

    static func toDomain(_ barcodes: [Barcode]) -> BarcodeScanningResult {
      return BarcodeScanningResult(barcodes: barcodes.map { toBarcodeData($0) })
    }

    private static func toBarcodeData(_ barcode: Barcode) -> BarcodeData {
      return BarcodeData(
        bounds: toBoundingBox(barcode.frame),
        corners: barcode.cornerPoints?.map { toCorner($0.cgPointValue) },
        format: Int(barcode.format.rawValue),
        formatName: formatName(barcode.format),
        valueType: Int(barcode.valueType.rawValue),
        valueTypeName: valueTypeName(barcode.valueType),
        rawValue: barcode.rawValue,
        displayValue: barcode.displayValue,
        rawBytes: barcode.rawData.map { Array($0).map { Int($0) } },
        isPotential: barcode.rawValue == nil && barcode.rawData == nil,
        value: toParsedValue(barcode)
      )
    }

    private static func toParsedValue(_ barcode: Barcode) -> BarcodeParsedValue? {
      if let wifi = barcode.wifi {
        return BarcodeParsedValue(
          type: "TYPE_WIFI",
          data: [
            "ssid": wifi.ssid,
            "password": wifi.password,
            "encryptionType": Int(wifi.type.rawValue),
            "encryptionTypeName": wifiEncryptionName(wifi.type),
          ]
        )
      }

      if let url = barcode.url {
        return BarcodeParsedValue(
          type: "TYPE_URL",
          data: [
            "title": url.title,
            "url": url.url,
          ]
        )
      }

      if let sms = barcode.sms {
        return BarcodeParsedValue(
          type: "TYPE_SMS",
          data: [
            "message": sms.message,
            "phoneNumber": sms.phoneNumber,
          ]
        )
      }

      if let email = barcode.email {
        return BarcodeParsedValue(
          type: "TYPE_EMAIL",
          data: [
            "address": email.address,
            "subject": email.subject,
            "body": email.body,
            "type": Int(email.type.rawValue),
            "typeName": emailTypeName(email.type),
          ]
        )
      }

      if let phone = barcode.phone {
        return BarcodeParsedValue(
          type: "TYPE_PHONE",
          data: [
            "number": phone.number,
            "type": Int(phone.type.rawValue),
            "typeName": phoneTypeName(phone.type),
          ]
        )
      }

      if let geo = barcode.geoPoint {
        return BarcodeParsedValue(
          type: "TYPE_GEO",
          data: [
            "lat": geo.latitude,
            "lng": geo.longitude,
          ]
        )
      }

      if let calendarEvent = barcode.calendarEvent {
        return BarcodeParsedValue(
          type: "TYPE_CALENDAR_EVENT",
          data: [
            "summary": calendarEvent.summary,
            "description": calendarEvent.eventDescription,
            "location": calendarEvent.location,
            "organizer": calendarEvent.organizer,
            "status": calendarEvent.status,
            "start": toDateInfo(calendarEvent.start),
            "end": toDateInfo(calendarEvent.end),
          ]
        )
      }

      if let contactInfo = barcode.contactInfo {
        return BarcodeParsedValue(
          type: "TYPE_CONTACT_INFO",
          data: [
            "name": contactInfo.name.map { toPersonName($0) },
            "organization": contactInfo.organization,
            "title": contactInfo.jobTitle,
            "phones": contactInfo.phones?.map { toPhone($0) },
            "emails": contactInfo.emails?.map { toEmail($0) },
            "addresses": contactInfo.addresses?.map { toAddress($0) },
            "urls": contactInfo.urls,
          ]
        )
      }

      if let driverLicense = barcode.driverLicense {
        return BarcodeParsedValue(
          type: "TYPE_DRIVER_LICENSE",
          data: [
            "documentType": driverLicense.documentType,
            "licenseNumber": driverLicense.licenseNumber,
            "firstName": driverLicense.firstName,
            "middleName": driverLicense.middleName,
            "lastName": driverLicense.lastName,
            "gender": driverLicense.gender,
            "addressStreet": driverLicense.addressStreet,
            "addressCity": driverLicense.addressCity,
            "addressState": driverLicense.addressState,
            "addressZip": driverLicense.addressZip,
            "birthDate": driverLicense.birthDate,
            "issueDate": driverLicense.issuingDate,
            "expiryDate": driverLicense.expiryDate,
            "issuingCountry": driverLicense.issuingCountry,
          ]
        )
      }

      if let rawValue = barcode.rawValue {
        let type = valueTypeName(barcode.valueType)
        let dataKey: String
        switch type {
        case "TYPE_TEXT": dataKey = "text"
        case "TYPE_PRODUCT": dataKey = "product"
        case "TYPE_ISBN": dataKey = "isbn"
        default: dataKey = "rawValue"
        }

        return BarcodeParsedValue(
          type: type,
          data: [dataKey: rawValue]
        )
      }

      return nil
    }

    private static func toPersonName(_ name: BarcodePersonName) -> [String: Any?] {
      return [
        "formattedName": name.formattedName,
        "pronunciation": name.pronunciation,
        "prefix": name.prefix,
        "first": name.first,
        "middle": name.middle,
        "last": name.last,
        "suffix": name.suffix,
      ]
    }

    private static func toPhone(_ phone: BarcodePhone) -> [String: Any?] {
      return [
        "number": phone.number,
        "type": Int(phone.type.rawValue),
        "typeName": phoneTypeName(phone.type),
      ]
    }

    private static func toEmail(_ email: BarcodeEmail) -> [String: Any?] {
      return [
        "address": email.address,
        "subject": email.subject,
        "body": email.body,
        "type": Int(email.type.rawValue),
        "typeName": emailTypeName(email.type),
      ]
    }

    private static func toAddress(_ address: BarcodeAddress) -> [String: Any?] {
      return [
        "addressLines": address.addressLines,
        "type": Int(address.type.rawValue),
        "typeName": addressTypeName(address.type),
      ]
    }

    private static func toDateInfo(_ date: Date?) -> [String: Any?]? {
      guard let date else {
        return nil
      }

      let calendar = Calendar(identifier: .gregorian)
      let components = calendar.dateComponents(
        [.year, .month, .day, .hour, .minute, .second],
        from: date
      )

      return [
        "year": components.year,
        "month": components.month,
        "day": components.day,
        "hours": components.hour,
        "minutes": components.minute,
        "seconds": components.second,
        "isUtc": false,
        "rawValue": ISO8601DateFormatter().string(from: date),
      ]
    }

    private static func toBoundingBox(_ rect: CGRect) -> DomainBoundingBox {
      let offsetX = (rect.midX - ceil(rect.width)) / 2.0
      let offsetY = (rect.midY - ceil(rect.height)) / 2.0

      let x = rect.maxX + offsetX
      let y = rect.minY + offsetY

      return DomainBoundingBox(
        x: rect.midX + (rect.midX - x),
        y: rect.midY + (y - rect.midY),
        centerX: rect.midX,
        centerY: rect.midY,
        width: rect.width,
        height: rect.height,
        top: rect.maxY,
        left: rect.minX,
        bottom: rect.minY,
        right: rect.maxX
      )
    }

    private static func toCorner(_ point: CGPoint) -> DomainCorner {
      return DomainCorner(x: Double(point.x), y: Double(point.y))
    }

    private static func formatName(_ format: MLKitBarcodeScanning.BarcodeFormat) -> String {
      switch Int(format.rawValue) {
      case 1: return "CODE_128"
      case 2: return "CODE_39"
      case 4: return "CODE_93"
      case 8: return "CODABAR"
      case 16: return "DATA_MATRIX"
      case 32: return "EAN_13"
      case 64: return "EAN_8"
      case 128: return "ITF"
      case 256: return "QR_CODE"
      case 512: return "UPC_A"
      case 1024: return "UPC_E"
      case 2048: return "PDF417"
      case 4096: return "AZTEC"
      default: return "UNKNOWN"
      }
    }

    private static func valueTypeName(_ valueType: MLKitBarcodeScanning.BarcodeValueType) -> String
    {
      switch Int(valueType.rawValue) {
      case 1: return "TYPE_CONTACT_INFO"
      case 2: return "TYPE_EMAIL"
      case 3: return "TYPE_ISBN"
      case 4: return "TYPE_PHONE"
      case 5: return "TYPE_PRODUCT"
      case 6: return "TYPE_SMS"
      case 7: return "TYPE_TEXT"
      case 8: return "TYPE_URL"
      case 9: return "TYPE_WIFI"
      case 10: return "TYPE_GEO"
      case 11: return "TYPE_CALENDAR_EVENT"
      case 12: return "TYPE_DRIVER_LICENSE"
      default: return "TYPE_UNKNOWN"
      }
    }

    private static func wifiEncryptionName(_ type: BarcodeWiFiEncryptionType) -> String {
      switch Int(type.rawValue) {
      case 1: return "OPEN"
      case 2: return "WPA"
      case 3: return "WEP"
      default: return "UNKNOWN"
      }
    }

    private static func phoneTypeName(_ type: BarcodePhoneType) -> String {
      switch Int(type.rawValue) {
      case 1: return "WORK"
      case 2: return "HOME"
      case 3: return "FAX"
      case 4: return "MOBILE"
      default: return "UNKNOWN"
      }
    }

    private static func emailTypeName(_ type: BarcodeEmailType) -> String {
      switch Int(type.rawValue) {
      case 1: return "WORK"
      case 2: return "HOME"
      default: return "UNKNOWN"
      }
    }

    private static func addressTypeName(_ type: BarcodeAddressType) -> String {
      switch Int(type.rawValue) {
      case 1: return "WORK"
      case 2: return "HOME"
      default: return "UNKNOWN"
      }
    }
  }
#endif  // MLKIT_BARCODE_SCANNING

File: android/src/main/java/com/visioncameramlkit/infrastructure/mlkit/adapters/MLKitBarcodeAdapter.kt
package com.visioncameramlkit.infrastructure.mlkit.adapters

import com.google.mlkit.vision.barcode.common.Barcode
import com.visioncameramlkit.domain.models.BarcodeData
import com.visioncameramlkit.domain.models.BarcodeParsedValue
import com.visioncameramlkit.domain.models.BarcodeScanningResult
import com.visioncameramlkit.domain.models.BoundingBox
import com.visioncameramlkit.domain.models.Corner

@Suppress("TooManyFunctions")
object MLKitBarcodeAdapter {
  private const val BYTE_MASK = 0xFF

  private val formatNameMap: Map<Int, String> =
    mapOf(
      Barcode.FORMAT_CODE_128 to "CODE_128",
      Barcode.FORMAT_CODE_39 to "CODE_39",
      Barcode.FORMAT_CODE_93 to "CODE_93",
      Barcode.FORMAT_CODABAR to "CODABAR",
      Barcode.FORMAT_DATA_MATRIX to "DATA_MATRIX",
      Barcode.FORMAT_EAN_13 to "EAN_13",
      Barcode.FORMAT_EAN_8 to "EAN_8",
      Barcode.FORMAT_ITF to "ITF",
      Barcode.FORMAT_QR_CODE to "QR_CODE",
      Barcode.FORMAT_UPC_A to "UPC_A",
      Barcode.FORMAT_UPC_E to "UPC_E",
      Barcode.FORMAT_PDF417 to "PDF417",
      Barcode.FORMAT_AZTEC to "AZTEC",
      Barcode.FORMAT_UNKNOWN to "UNKNOWN",
    )

  private val valueTypeNameMap: Map<Int, String> =
    mapOf(
      Barcode.TYPE_CONTACT_INFO to "TYPE_CONTACT_INFO",
      Barcode.TYPE_EMAIL to "TYPE_EMAIL",
      Barcode.TYPE_ISBN to "TYPE_ISBN",
      Barcode.TYPE_PHONE to "TYPE_PHONE",
      Barcode.TYPE_PRODUCT to "TYPE_PRODUCT",
      Barcode.TYPE_SMS to "TYPE_SMS",
      Barcode.TYPE_TEXT to "TYPE_TEXT",
      Barcode.TYPE_URL to "TYPE_URL",
      Barcode.TYPE_WIFI to "TYPE_WIFI",
      Barcode.TYPE_GEO to "TYPE_GEO",
      Barcode.TYPE_CALENDAR_EVENT to "TYPE_CALENDAR_EVENT",
      Barcode.TYPE_DRIVER_LICENSE to "TYPE_DRIVER_LICENSE",
      Barcode.TYPE_UNKNOWN to "TYPE_UNKNOWN",
    )

  private val parsedValueExtractors: Map<Int, (Barcode) -> BarcodeParsedValue?> =
    mapOf(
      Barcode.TYPE_WIFI to { barcode -> barcode.wifi?.let(::toWifi) },
      Barcode.TYPE_URL to { barcode -> barcode.url?.let(::toUrl) },
      Barcode.TYPE_SMS to { barcode -> barcode.sms?.let(::toSms) },
      Barcode.TYPE_EMAIL to { barcode -> barcode.email?.let(::toEmail) },
      Barcode.TYPE_PHONE to { barcode -> barcode.phone?.let(::toPhone) },
      Barcode.TYPE_CONTACT_INFO to { barcode -> barcode.contactInfo?.let(::toContactInfo) },
      Barcode.TYPE_CALENDAR_EVENT to { barcode -> barcode.calendarEvent?.let(::toCalendarEvent) },
      Barcode.TYPE_GEO to { barcode -> barcode.geoPoint?.let(::toGeoPoint) },
      Barcode.TYPE_DRIVER_LICENSE to { barcode -> barcode.driverLicense?.let(::toDriverLicense) },
    )

  private val rawValueTypeMappings: Map<Int, Pair<String, String>> =
    mapOf(
      Barcode.TYPE_TEXT to ("TYPE_TEXT" to "text"),
      Barcode.TYPE_PRODUCT to ("TYPE_PRODUCT" to "product"),
      Barcode.TYPE_ISBN to ("TYPE_ISBN" to "isbn"),
    )

  fun toDomain(barcodes: List<Barcode>): BarcodeScanningResult =
    BarcodeScanningResult(
      barcodes = barcodes.map(::toBarcodeData),
    )

  private fun toBarcodeData(barcode: Barcode): BarcodeData =
    BarcodeData(
      bounds = barcode.boundingBox?.let(::toBoundingBox),
      corners = barcode.cornerPoints?.map(::toCorner),
      format = barcode.format,
      formatName = formatName(barcode.format),
      valueType = barcode.valueType,
      valueTypeName = valueTypeName(barcode.valueType),
      rawValue = barcode.rawValue,
      displayValue = barcode.displayValue,
      rawBytes = barcode.rawBytes?.map { byte -> byte.toInt() and BYTE_MASK },
      isPotential = barcode.rawValue == null && barcode.rawBytes == null,
      value = toParsedValue(barcode),
    )

  private fun toParsedValue(barcode: Barcode): BarcodeParsedValue? {
    parsedValueExtractors[barcode.valueType]?.invoke(barcode)?.let { parsed ->
      return parsed
    }

    val rawMapping = rawValueTypeMappings[barcode.valueType] ?: return null
    val rawValue = barcode.rawValue ?: return null
    return BarcodeParsedValue(
      type = rawMapping.first,
      data = mapOf(rawMapping.second to rawValue),
    )
  }

  private fun toWifi(wifi: Barcode.WiFi): BarcodeParsedValue =
    BarcodeParsedValue(
      type = "TYPE_WIFI",
      data =
        mapOf(
          "ssid" to wifi.ssid,
          "password" to wifi.password,
          "encryptionType" to wifi.encryptionType,
          "encryptionTypeName" to wifiEncryptionName(wifi.encryptionType),
        ),
    )

  private fun toUrl(url: Barcode.UrlBookmark): BarcodeParsedValue =
    BarcodeParsedValue(
      type = "TYPE_URL",
      data =
        mapOf(
          "title" to url.title,
          "url" to url.url,
        ),
    )

  private fun toSms(sms: Barcode.Sms): BarcodeParsedValue =
    BarcodeParsedValue(
      type = "TYPE_SMS",
      data =
        mapOf(
          "message" to sms.message,
          "phoneNumber" to sms.phoneNumber,
        ),
    )

  private fun toEmail(email: Barcode.Email): BarcodeParsedValue =
    BarcodeParsedValue(
      type = "TYPE_EMAIL",
      data =
        mapOf(
          "address" to email.address,
          "subject" to email.subject,
          "body" to email.body,
          "type" to email.type,
          "typeName" to emailTypeName(email.type),
        ),
    )

  private fun toPhone(phone: Barcode.Phone): BarcodeParsedValue =
    BarcodeParsedValue(
      type = "TYPE_PHONE",
      data =
        mapOf(
          "number" to phone.number,
          "type" to phone.type,
          "typeName" to phoneTypeName(phone.type),
        ),
    )

  private fun toContactInfo(contactInfo: Barcode.ContactInfo): BarcodeParsedValue =
    BarcodeParsedValue(
      type = "TYPE_CONTACT_INFO",
      data =
        mapOf(
          "name" to contactInfo.name?.let(::toPersonName),
          "organization" to contactInfo.organization,
          "title" to contactInfo.title,
          "phones" to contactInfo.phones.map(::toPhoneMap),
          "emails" to contactInfo.emails.map(::toEmailMap),
          "addresses" to contactInfo.addresses.map(::toAddressMap),
          "urls" to contactInfo.urls,
        ),
    )

  private fun toCalendarEvent(event: Barcode.CalendarEvent): BarcodeParsedValue =
    BarcodeParsedValue(
      type = "TYPE_CALENDAR_EVENT",
      data =
        mapOf(
          "summary" to event.summary,
          "description" to event.description,
          "location" to event.location,
          "organizer" to event.organizer,
          "status" to event.status,
          "start" to event.start?.let(::toCalendarDateTime),
          "end" to event.end?.let(::toCalendarDateTime),
        ),
    )

  private fun toGeoPoint(geoPoint: Barcode.GeoPoint): BarcodeParsedValue =
    BarcodeParsedValue(
      type = "TYPE_GEO",
      data =
        mapOf(
          "lat" to geoPoint.lat,
          "lng" to geoPoint.lng,
        ),
    )

  private fun toDriverLicense(driverLicense: Barcode.DriverLicense): BarcodeParsedValue =
    BarcodeParsedValue(
      type = "TYPE_DRIVER_LICENSE",
      data =
        mapOf(
          "documentType" to driverLicense.documentType,
          "licenseNumber" to driverLicense.licenseNumber,
          "firstName" to driverLicense.firstName,
          "middleName" to driverLicense.middleName,
          "lastName" to driverLicense.lastName,
          "gender" to driverLicense.gender,
          "addressStreet" to driverLicense.addressStreet,
          "addressCity" to driverLicense.addressCity,
          "addressState" to driverLicense.addressState,
          "addressZip" to driverLicense.addressZip,
          "birthDate" to driverLicense.birthDate,
          "issueDate" to driverLicense.issueDate,
          "expiryDate" to driverLicense.expiryDate,
          "issuingCountry" to driverLicense.issuingCountry,
        ),
    )

  private fun toPersonName(name: Barcode.PersonName): Map<String, Any?> =
    mapOf(
      "formattedName" to name.formattedName,
      "pronunciation" to name.pronunciation,
      "prefix" to name.prefix,
      "first" to name.first,
      "middle" to name.middle,
      "last" to name.last,
      "suffix" to name.suffix,
    )

  private fun toPhoneMap(phone: Barcode.Phone): Map<String, Any?> =
    mapOf(
      "number" to phone.number,
      "type" to phone.type,
      "typeName" to phoneTypeName(phone.type),
    )

  private fun toEmailMap(email: Barcode.Email): Map<String, Any?> =
    mapOf(
      "address" to email.address,
      "subject" to email.subject,
      "body" to email.body,
      "type" to email.type,
      "typeName" to emailTypeName(email.type),
    )

  private fun toAddressMap(address: Barcode.Address): Map<String, Any?> =
    mapOf(
      "addressLines" to address.addressLines.toList(),
      "type" to address.type,
      "typeName" to addressTypeName(address.type),
    )

  private fun toCalendarDateTime(dateTime: Barcode.CalendarDateTime): Map<String, Any?> =
    mapOf(
      "year" to dateTime.year,
      "month" to dateTime.month,
      "day" to dateTime.day,
      "hours" to dateTime.hours,
      "minutes" to dateTime.minutes,
      "seconds" to dateTime.seconds,
      "isUtc" to dateTime.isUtc,
      "rawValue" to dateTime.rawValue,
    )

  private fun toBoundingBox(rect: android.graphics.Rect): BoundingBox =
    BoundingBox(
      x = rect.exactCenterX().toDouble(),
      y = rect.exactCenterY().toDouble(),
      centerX = rect.centerX().toDouble(),
      centerY = rect.centerY().toDouble(),
      width = rect.width().toDouble(),
      height = rect.height().toDouble(),
      top = rect.top.toDouble(),
      left = rect.left.toDouble(),
      bottom = rect.bottom.toDouble(),
      right = rect.right.toDouble(),
    )

  private fun toCorner(point: android.graphics.Point): Corner =
    Corner(
      x = point.x.toDouble(),
      y = point.y.toDouble(),
    )

  private fun formatName(format: Int): String = formatNameMap[format] ?: "UNKNOWN"

  private fun valueTypeName(valueType: Int): String = valueTypeNameMap[valueType] ?: "TYPE_UNKNOWN"

  private fun wifiEncryptionName(type: Int): String =
    when (type) {
      Barcode.WiFi.TYPE_OPEN -> "OPEN"
      Barcode.WiFi.TYPE_WEP -> "WEP"
      Barcode.WiFi.TYPE_WPA -> "WPA"
      else -> "UNKNOWN"
    }

  private fun phoneTypeName(type: Int): String =
    when (type) {
      Barcode.Phone.TYPE_FAX -> "FAX"
      Barcode.Phone.TYPE_HOME -> "HOME"
      Barcode.Phone.TYPE_MOBILE -> "MOBILE"
      Barcode.Phone.TYPE_WORK -> "WORK"
      else -> "UNKNOWN"
    }

  private fun emailTypeName(type: Int): String =
    when (type) {
      Barcode.Email.TYPE_HOME -> "HOME"
      Barcode.Email.TYPE_WORK -> "WORK"
      else -> "UNKNOWN"
    }

  private fun addressTypeName(type: Int): String =
    when (type) {
      Barcode.Address.TYPE_HOME -> "HOME"
      Barcode.Address.TYPE_WORK -> "WORK"
      else -> "UNKNOWN"
    }
}
